cachi2/core/package_managers/generic/models.py

```diff
@@ -29,9 +42,51 @@ def no_empty_checksums(cls, value: dict[str, str]) -> dict[str, str]:
             raise ValueError("At least one checksum must be provided.")
         return value
 
+    @model_validator(mode="after")
+    def set_target(self, info: ValidationInfo) -> "LockfileArtifact":
+        """Set the target path if not provided and resolve it into an absolute path."""
+        if not self.target:
+            url_path = urlparse(str(self.download_url)).path
+            self.target = Path(url_path).name
+
+        # needs context
+        if not info.context or "output_dir" not in info.context:
+            raise PackageManagerError(
+                "The `LockfileArtifact` class needs to be called with `output_dir` in the context"
+            )
+        output_dir: RootedPath = info.context["output_dir"]
+        self.target = str(output_dir.join_within_root(self.target).path.resolve())
+
+        return self
+
+    @property
+    def formatted_checksums(self) -> List[ChecksumInfo]:
+        """Return the checksums as a list of ChecksumInfo objects."""
+        return [ChecksumInfo(algo, digest) for algo, digest in self.checksums.items()]
+
 
 class GenericLockfileV1(BaseModel):
     """Defines format of the cachi2 generic lockfile, version 1.0."""
 
     metadata: LockfileMetadata
     artifacts: list[LockfileArtifact]
+
+    @model_validator(mode="after")
+    def no_artifact_conflicts(self) -> "GenericLockfileV1":
+        """Validate that all artifacts have unique targets and download_urls."""
+        urls = set()
+        targets = set()
```

Comment:
Do these need to be sets ?

Reply:
Why not?

Reply:
Because you are checking if something already exists in the set. That seems a bit off to me. Sets do not contain duplicate values. That's why I raised the question.

Reply:
Because sets have O(1) complexity on lookup which is much faster than searching a list (O(N)) in the loop below.